File: src/anomalies.py
from typing import Optional
from sqlalchemy.orm import Session
from sqlalchemy import and_
from collector.models import Weather, Anomaly, LocationToTrack
from db import get_session
from src.transform import collect_weather


db = get_session()


def compare_weather(db: Session, structured_weather: Weather) -> Optional[dict]:
    lat = structured_weather.lat
    lon = structured_weather.lon
    temperature = structured_weather.temperature
    pressure = structured_weather.pressure
    humidity = structured_weather.humidity
    wind_speed = structured_weather.wind_speed

    similar_places = []
    if lon is not None and lat is not None:
        print("Ищем похожие места")
        radius = 1.0
        similar_places = db.query(Weather).filter(
            and_(  # ← 4 УСЛОВИЯ and_, НЕ or_!
                Weather.lat >= lat - radius,
                Weather.lat <= lat + radius,
                Weather.lon >= lon - radius,
                Weather.lon <= lon + radius
            )
        ).all()
        print(f"Найдено похожих: {len(similar_places)} в радиусе {radius}° от ({lat}, {lon})")
    if len(similar_places) < 2:
        print("Похожие места не найдены или их не хватает")
        return None

    temps = [place.temperature for place in similar_places if place.temperature is not None]
    pressures = [place.pressure for place in similar_places if place.pressure is not None]
    humidities = [place.humidity for place in similar_places if place.humidity is not None]
    winds = [place.wind_speed for place in similar_places if place.wind_speed is not None]

    if not temps:
        return None

    avg_temp = sum(temps) / len(temps)
    avg_pressure = sum(pressures) / len(pressures)
    avg_humidity = sum(humidities) / len(humidities)
    avg_wind = sum(winds) / len(winds)

    threshold = 0.3
    anomaly = {
            'temperature': abs(temperature - avg_temp) > threshold * abs(avg_temp) if temperature else False,
            'pressure': abs(pressure - avg_pressure) > threshold * abs(avg_pressure) if pressure else False,
            'humidity': abs(humidity - avg_humidity) > threshold * avg_humidity if humidity else False,
            'wind_speed': abs(wind_speed - avg_wind) > threshold * avg_wind if wind_speed else False,
            'averages': {'temperature': avg_temp, 'pressure': avg_pressure, 'humidity': avg_humidity, 'wind_speed': avg_wind}
    }
    return anomaly


def check_anomalies(db: Session, lat: float, lon: float, loc: LocationToTrack) -> Optional[Anomaly]:
    weather = collect_weather(db, lat=lat, lon=lon)
    if weather:
        print("Проверка аномалий")
        anomalies = compare_weather(db, weather)
        if anomalies:

            # Собираем все аномалии для этой локации
            anomalies_to_save = {}
            anomalies_data = {}

            for key, is_anomaly in anomalies.items():
                if is_anomaly and key != 'averages':
                    anomaly_value = getattr(weather, key)
                    print(f"Аномалия: {key}: {anomaly_value}! Среднее: {anomalies['averages'][key]}")

                    # Определяем правильное название поля для БД
                    if key == "temperature":
                        anomalies_to_save['anomaly_temperature'] = anomaly_value
                    elif key == "pressure":
                        anomalies_to_save['anomaly_pressure'] = anomaly_value
                    elif key == "humidity":
                        anomalies_to_save['anomaly_humidity'] = anomaly_value
                    elif key == "wind_speed":
                        anomalies_to_save['anomaly_wind_speed'] = anomaly_value

                    # Сохраняем данные для additional_data
                    anomalies_data[key] = {
                        "value": anomaly_value,
                        "avg": anomalies['averages'][key]
                    }

                else:
                    if key != 'averages':
                        value = getattr(weather, key)
                        print(f"Аномалий в {key} НЕ найдено! Среднее: {anomalies['averages'][key]}, текущее: {value}")

            if anomalies_to_save:
                saved_anomaly = save_anomaly(db, anomalies_to_save, loc, anomalies_data)
                return saved_anomaly

    return None


def save_anomaly(db: Session, anomalies: dict, loc: LocationToTrack, data: dict) -> Optional[Anomaly]:
    db_anomaly = Anomaly(
        location=loc,
        additional_data=data,
        **anomalies  # Распаковываем все аномалии в конструктор
    )
    db.add(db_anomaly)
    db.commit()
    db.refresh(db_anomaly)
    return db_anomaly

File: golang/analyze/anomalies.go
package analyze

import (
	"context"
	"fmt"
	"log"
	"math"
	"os"
	"time"
	"github.com/jackc/pgx/v5/pgxpool"
)

type Weather struct {
	Temperature *float64
	Pressure    *float64
	Humidity    *float64
	WindSpeed   *float64
}

type APIWeather struct {
	ID          int       `json:"id"`
	LocId       int       `json:"loc_id"`
	Lat         float64   `json:"lat"`
	Lon         float64   `json:"lon"`
	Temperature *float64  `json:"temperature"`
	Pressure    *float64  `json:"pressure"`
	Humidity    *float64  `json:"humidity"`
	WindSpeed   *float64  `json:"wind_speed"`
	CollectedAt time.Time `json:"collected_at"`
}

type ErrorResponse struct {
	Error string `json:"error"`
}


func Sum[T int | float64](slice []T) T {
	var result T
	for _, value := range slice {
		result += value
	}
	return result
}

func ConnectToPostgres() *pgxpool.Pool {
	ctx := context.Background()
	dsn := os.Getenv("DATABASE_URL_FOR_GO")
	if dsn == "" {
		dsn = "postgresql://weather_user:weather_pass@weather_db:5432/weather_guard"
	}

	pool, err := pgxpool.New(ctx, dsn)
	if err != nil {
		log.Fatal(err)
	}
	// defer pool.Close()

	if err := pool.Ping(ctx); err != nil {
		log.Fatal(err)
	}

	fmt.Println("Connected to Postgres")
	return pool
}
func isAnomaly(val *float64, avg, threshold float64, useAbsAvg bool) bool {
	if val == nil {
		return false
	}
	diff := math.Abs(*val - avg)
	if useAbsAvg {
		return diff > threshold*math.Abs(avg)
	}
	return diff > threshold*avg
}

func compareWeather(pool *pgxpool.Pool, ctx context.Context, data APIWeather) map[string]any {
	lat := data.Lat
	lon := data.Lon
	temperature := data.Temperature
	pressure := data.Pressure
	humidity := data.Humidity
	windSpeed := data.WindSpeed

	radius := 0.3

	similarPlacesRows, err := pool.Query(ctx,
		`SELECT temperature, pressure, humidity, wind_speed FROM weather
		WHERE lat >= $1 AND lat <= $2 AND lon >= $3 AND lon <= $4 AND collected_at > $5`,
		lat-radius, lat+radius, lon-radius, lon+radius, time.Now().UTC().AddDate(0, 0, -31))
	if err != nil {
		log.Fatal(err)
	}
	defer similarPlacesRows.Close()

	var similarPlaces []Weather
	for similarPlacesRows.Next() {
		var w Weather
		if err := similarPlacesRows.Scan(&w.Temperature, &w.Pressure, &w.Humidity, &w.WindSpeed); err != nil {
			log.Fatal(err)
		}
		similarPlaces = append(similarPlaces, w)
	}
	if similarPlacesRows.Err() != nil {
		log.Fatal(similarPlacesRows.Err())
	}
	if len(similarPlaces) < 2 {
		fmt.Println("Похожие места не найдены или их не хватает")
		return nil
	}

	var temps []float64
	var pressures []float64
	var humidities []float64
	var winds []float64
	for _, place := range similarPlaces {
		if place.Temperature != nil {
			temps = append(temps, *place.Temperature)
		}
		if place.Humidity != nil {
			humidities = append(humidities, *place.Humidity)
		}
		if place.Pressure != nil {
			pressures = append(pressures, *place.Pressure)
		}
		if place.WindSpeed != nil {
			winds = append(winds, *place.WindSpeed)
		}
	}
	var avgTemp, avgPressure, avgHumidity, avgWind float64
	if len(temps) > 0 {avgTemp = Sum(temps) / float64(len(temps))}else {avgTemp = 0}
	if len(humidities) > 0 {avgHumidity = Sum(humidities) / float64(len(humidities))}else {avgHumidity = 0}
	if len(pressures) > 0 {avgPressure= Sum(pressures) / float64(len(pressures))}else {avgPressure = 0}
	if len(winds) > 0 {avgWind = Sum(winds) / float64(len(winds))}else {avgWind = 0}

	threshold := 0.65
	anomaly := map[string]any{
		"temperature": isAnomaly(temperature, avgTemp, threshold, true),
		"pressure":    isAnomaly(pressure, avgPressure, threshold, true),
		"humidity":    isAnomaly(humidity, avgHumidity, threshold, false),
		"wind_speed":  isAnomaly(windSpeed, avgWind, threshold, false),
		"averages": map[string]float64{
			"temperature": avgTemp,
			"pressure":    avgPressure,
			"humidity":    avgHumidity,
			"wind_speed":  avgWind,
		},
	}
	return anomaly

}

func getField(w APIWeather, field string) *float64 {
	switch field {
	case "temperature":
		return w.Temperature
	case "humidity":
		return w.Humidity
	case "pressure":
		return w.Pressure
	case "wind_speed":
		return w.WindSpeed
	default:
		return nil
	}
}

func CheckAnomalies(pool *pgxpool.Pool, ctx context.Context, data APIWeather) map[string]any {
	anomalies := compareWeather(pool, ctx, data)
	if anomalies != nil {
		var anomalies_to_save map[string]*float64 = make(map[string]*float64)
		var anomalies_data map[string]map[string]any = make(map[string]map[string]any)

		for key, is_anomaly := range anomalies {
			if is_anomaly == true && key != "averages" {
				anomaly_value := getField(data, key)
				if anomaly_value != nil {
					fmt.Printf("Аномалия: %s: %v! Среднее: %f\n", key, anomaly_value, anomalies["averages"].(map[string]float64)[key])

					switch key {
					case "temperature":
						anomalies_to_save["anomaly_temperature"] = anomaly_value
					case "pressure":
						anomalies_to_save["anomaly_pressure"] = anomaly_value
					case "humidity":
						anomalies_to_save["anomaly_humidity"] = anomaly_value
					case "wind_speed":
						anomalies_to_save["anomaly_wind_speed"] = anomaly_value
					}

					anomalies_data[key] = map[string]any{"value": anomaly_value, "avg": anomalies["averages"].(map[string]float64)[key]}
				}
			}
		}

		if len(anomalies_to_save) > 0 {
			// saved_anomaly := saveAnomaly(pool, ctx, anomalies_to_save, data.LocId, anomalies_data)
			result := map[string]any{
				"loc_id":            data.LocId,
				"anomalies_to_save": anomalies_to_save,
				"anomalies_data":    anomalies_data,
			}
			return result
		}
	}

	return nil
}